recurly.module

```diff
@@ -914,202 +913,28 @@ function recurly_format_coupon(\Recurly_Coupon $coupon, $currency, $html = FALSE
-    $variables['currency'] = $currency;
-  }
-
-  // Prepare an easy to loop-through list of subscriptions.
-  $variables['filtered_plans'] = [];
-  foreach ($plans as $plan_code => $plan) {
-    $setup_fee_amount = NULL;
-    foreach ($plan->setup_fee_in_cents as $setup_currency) {
-      if ($setup_currency->currencyCode === $currency) {
-        $setup_fee_amount = recurly_format_currency($setup_currency->amount_in_cents, $setup_currency->currencyCode, TRUE);
-        break;
-      }
-    }
-    $unit_amount = NULL;
-    foreach ($plan->unit_amount_in_cents as $unit_currency) {
-      if ($unit_currency->currencyCode === $currency) {
-        $unit_amount = recurly_format_currency($unit_currency->amount_in_cents, $unit_currency->currencyCode, TRUE);
-        break;
-      }
-    }
-    $variables['filtered_plans'][$plan_code] = [
-      'plan_code' => SafeMarkup::checkPlain($plan_code),
-      'name' => SafeMarkup::checkPlain($plan->name),
-      'description' => SafeMarkup::checkPlain($plan->description),
-      'setup_fee' => $setup_fee_amount,
-      'amount' => $unit_amount,
-      'plan_interval' => recurly_format_price_interval($unit_amount, $plan->plan_interval_length, $plan->plan_interval_unit, TRUE),
-      'trial_interval' => $plan->trial_interval_length ? recurly_format_price_interval(NULL, $plan->trial_interval_length, $plan->trial_interval_unit, TRUE) : NULL,
-      'signup_url' => recurly_url('subscribe', [
-        'entity_type' => $entity_type,
-        'entity' => $entity,
-        'plan_code' => $plan_code,
-        'currency' => $currency,
-      ]),
-      'change_url' => $current_subscription ? recurly_url('change_plan', [
-        'entity_type' => $entity_type,
-        'entity' => $entity,
-        'subscription' => $current_subscription,
-        'plan_code' => $plan_code,
-      ]) : NULL,
-      'selected' => FALSE,
-    ];
-
-    // If we have a pending subscription, make that its shown as selected rather
-    // than the current active subscription. This should allow users to switch
-    // back to a previous plan after making a pending switch to another one.
-    foreach ($subscriptions as $subscription) {
-      if (!empty($subscription->pending_subscription)) {
-        if ($subscription->pending_subscription->plan->plan_code === $plan_code) {
-          $variables['filtered_plans'][$plan_code]['selected'] = TRUE;
-        }
-      }
-      elseif ($subscription->plan->plan_code === $plan_code) {
-        $variables['filtered_plans'][$plan_code]['selected'] = TRUE;
-      }
-    }
-  }
-
-  // Check if this is an account that is creating a new subscription.
-  $variables['expired_subscriptions'] = FALSE;
-  $account = recurly_account_load(['entity_type' => $entity_type, 'entity_id' => $entity->id()]);
-  if ($account) {
-    $variables['expired_subscriptions'] = empty($subscriptions);
-  }
+  \Drupal::service('recurly.recurly_preprocess')->preprocessRecurlySubscriptionPlanSelect($variables);
```

Comment:
From the reading I just did, it seems services have some "magic" (to me at the moment at least) way of handling all the dependencies they need when they are defined in YAML and injecting them. So that frees the developer from having to handle altering parameters and dealing with all that manually.